src/librustdoc/html/highlight.rs

```diff
@@ -134,151 +135,321 @@ fn can_merge(class1: Option<Class>, class2: Option<Class>, text: &str) -> bool {
+            } else {
+                out.write_str("</span>").unwrap();
+            }
+        }
+    }
+}
+
+/// This represents the stack of HTML elements. For example a macro expansion
+/// will contain other elements which might themselves contain other elements
+/// (like macros).
+///
+/// This allows to easily handle HTML tags instead of having a more complicated
+/// state machine to keep track of which tags are open.
+#[derive(Debug)]
+struct ClassStack {
+    open_classes: Vec<ClassInfo>,
+}
+
+impl ClassStack {
+    fn new() -> Self {
+        Self { open_classes: Vec::new() }
+    }
+
+    fn enter_elem<W: Write>(
+        &mut self,
+        out: &mut W,
+        href_context: &Option<HrefContext<'_, '_>>,
+        new_class: Class,
+        pending_exit: bool,
+    ) {
+        if let Some(current_class) = self.open_classes.last_mut() {
+            if can_merge(Some(current_class.class), Some(new_class), "") {
+                current_class.pending_exit = false;
+                return;
+            } else if current_class.pending_exit {
+                current_class.close_tag(out);
+                self.open_classes.pop();
+            }
+        }
+        let mut class_info = ClassInfo::new(new_class, pending_exit);
+        if pending_exit {
+            class_info.open = true;
+        } else if matches!(new_class, Class::Decoration(_) | Class::Original) {
+            // We open it right away to ensure it always come at the expected location.
+            // FIXME: Should we instead add a new boolean field to `ClassInfo` to force a non-open
+            // tags to be added if another one comes before it's open?
+            write!(out, "<span class=\"{}\">", new_class.as_html()).unwrap();
+            class_info.open = true;
+        } else if new_class.get_span().is_some()
+            && let Some(closing_tag) =
+                string_without_closing_tag(out, "", Some(class_info.class), href_context, false)
+            && !closing_tag.is_empty()
+        {
+            class_info.open = true;
+            class_info.link_closing_tag = closing_tag == "</a>";
+        }
+
+        self.open_classes.push(class_info);
+    }
+
+    /// This sets the `pending_exit` field to `true`. Meaning that if we try to push another stack
+    /// which is not compatible with this one, it will exit the current one before adding the new
+    /// one.
+    fn exit_elem(&mut self) {
+        let current_class =
+            self.open_classes.last_mut().expect("`exit_elem` called on empty class stack");
+        if !current_class.pending_exit {
+            current_class.pending_exit = true;
+            return;
+        }
+        // If the current class was already closed, it means we are actually closing its parent.
+        self.open_classes.pop();
+        let current_class =
+            self.open_classes.last_mut().expect("`exit_elem` called on empty class stack parent");
+        current_class.pending_exit = true;
```

Comment:
doesn't `pending_exit` merely signal that the closing element should be written, not that it has been written?  should we add an assertion that `open` is false for the popped class? otherwise we would be opening tags and then not closing them? am I missing something?